### PR TITLE
BUG: Fix IntervalIndex.get_loc/get_indexer for IntervalIndex of length one

### LINKED ISSUE #17284: qcut: Using cut with IntervalIndex provided by qcut producing wrong NaN values
xref #17282 

#### Code Sample

```python
>>> x.isnull().sum()
0

>>> x.value_counts()
0.000000     693
12.561725      1
13.568112      1
12.521249      1
13.007628      1
6.993961       1
14.815512      1
6.017280       1
12.944714      1
Name: 0, dtype: int64

>>> categorized = pd.qcut(x, 10, duplicates='drop')
>>> categorized.isnull().sum()
0

>>> categorized.cat.categories  # Notice how all values of x are contained in the only interval
IntervalIndex([(-0.001, 14.816]]
              closed='right',
              dtype='interval[float64]')

>>> res = pd.cut(x, categorized.cat.categories)
>>> res.isnull().sum()
701
```
Copy pastable

```python
x = pd.read_csv('x.csv', header=None).iloc[:, 0]  # x.csv is provided in a comment below
categorized = pd.qcut(x, 10, duplicates='drop')
res = pd.cut(x, categorized.cat.categories)
res.isnull().sum()
```
#### Problem description

When I use `qcut` to get the `IntervalIndex` corresponding to the quantiles of a `float64` series, and than use this as the `bins` of `cut` on the same float64 series, it doesn't work. It produces a new series with a lot of NaN values, while the original series contained no NaN and all of its values are contained at the interval of `IntervalIndex`.

#### Expected Output

The result of both `qcut` and `cut` should also be the same, but they are not.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-514.26.2.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.1
pytest: 3.0.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 5.3.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.3.0
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #20921: BUG: IntervalIndex.get_loc fails when there is only one entry
#### Code Sample, a copy-pastable example if possible

```python
>>> import pandas as pd
>>> pd.IntervalIndex.from_tuples([(1,100)]).get_loc(50)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/imre/code/pandas/pandas/core/indexes/interval.py", line 1049, in get_loc
    raise KeyError(original_key)
KeyError: 50
```
#### Problem description
50 is contained in the interval (1, 100), so this should not raise `KeyError`.

#### Expected Output
0

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: c4da79b5b322c73d8e61d1cb98ac4ab1e2438b40
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.0-39-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.0.dev0+824.gc4da79b5b
pytest: None
pip: 10.0.1
setuptools: 39.1.0
Cython: 0.28.2
numpy: 1.14.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by jschendel**
- [X] closes #17284
- [X] closes #20921
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -1245,6 +1245,7 @@ Indexing
 - Bug in ``Series.is_unique`` where extraneous output in stderr is shown if Series contains objects with ``__ne__`` defined (:issue:`20661`)
 - Bug in ``.loc`` assignment with a single-element list-like incorrectly assigns as a list (:issue:`19474`)
 - Bug in partial string indexing on a ``Series/DataFrame`` with a monotonic decreasing ``DatetimeIndex`` (:issue:`19362`)
+- Bug in ``IntervalIndex.get_loc()`` and ``IntervalIndex.get_indexer()`` when used with an :class:`IntervalIndex` containing a single interval (:issue:`17284`, :issue:`20921`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: are these i the api.rst? if not can you can add and put refs here?

File: pandas/tests/indexes/interval/test_interval.py [hunk 2]
```diff
@@ -544,6 +552,16 @@ def test_get_indexer_subintervals(self):
         expected = np.array([0, 0, 0], dtype='intp')
         tm.assert_numpy_array_equal(actual, expected)
 
+    # Make consistent with test_interval_new.py (see #16316, #16386)
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval.py]: if this is consistent with those tests, you can remove from there I think (not sure exactly how we are tracking 'new' things vs current)

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.0.txt (hunk 1 above)
Comment (jschendel) [reply doc/source/whatsnew/v0.23.0.txt]: done

File: pandas/tests/indexes/interval/test_interval.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/indexes/interval/test_interval.py]: k that is fine. yeah converting to use the new tests is a major effort :>

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jschendel. always nice to have a great patch from you!

**Pull request merged**